src/Microsoft.AspNetCore.Hosting/WebHostExtensions.cs

```diff
@@ -16,18 +19,27 @@ public static class WebHostExtensions
         /// <param name="host">The <see cref="IWebHost"/> to run.</param>
         public static void Run(this IWebHost host)
         {
+            var done = new ManualResetEvent(false);
             using (var cts = new CancellationTokenSource())
             {
-                Console.CancelKeyPress += (sender, eventArgs) =>
+                Action shutdown = () =>
                 {
-                    Console.WriteLine("Application is shutting down...");
-                    cts.Cancel();
+                    if (!cts.IsCancellationRequested)
+                    {
+                        Console.WriteLine("Application is shutting down...");
+                        cts.Cancel();
+                    }
 
-                    // Don't terminate the process immediately, wait for the Main thread to exit gracefully.
-                    eventArgs.Cancel = true;
+                    done.WaitOne();
```

Comment:
There is no way we can support this with SIGTERM. So we need to decide if we want to have two behaviors or one common.